### ISSUE TITLE
Redis / MongoDB persistence?

### ISSUE THREAD

**Issue by testark** [vernemq/vernemq#217]
Could you please shed some light on what the high-level steps would be in order to switch persistence of everything (queues, subscriptions, offline messages, etc.) from node-local LevelDB to an external Redis / MongoDB instance? Would this be something you would consider adding to VerneMQ (or perhaps develop as a custom-tailored, maybe paid plugin)?

**Comment by dergraf** [comment]
Sure.

It's something we have in mind and it's partially available for messages already today. However the `vmq_queue` has to be changed in a way that the queue migration phase is optional, as message migration is not required for a clusterable message backends (such as e.g. Riak).

With the commit https://github.com/erlio/vmq_server/commit/f1821d4468e19e6b828e7d4f00ee18856b12e1d3 we added a module `vmq_subscriber_db` that is the facade for interacting with the database storing the subscriber data. While it't a nice refactoring anyway it also enables to easily implement our current plumtree based storage as a plugin and just call the hooks inside the `vmq_subscriber_db`. The main reason why we didn't do that already is that our async queue migrations as well as the routing tables updates are triggered through events coming from the database. In order to have this pluggable we have to come up with a generic way to trigger those events. 

If it should be a solution that should work rather today than tomorrow I guess this would be a custom tailored thing, which would involve changing vmq_server. However, I would favor a generic thing, which could be opensourced of course. In this regard we're of course open to sponsoring! :)

**Comment by dcy** [comment]
Looking forward to it.
Does it mean that it can allow  clients to connect other nodes when a node ungracefully leaves the cluster If there is a clusterable message backend? 
https://github.com/erlio/vernemq/issues/193#issuecomment-243371976

**Comment by ig0r** [comment]
Guys,

Is there any news on this item? 

It'd be more scalable and reliable than a password file. Something similar to this [emq-auth-mongo](https://github.com/emqtt/emq-auth-mongo)

Thanks!

**Comment by larshesel** [comment]
@ig0r this issue is about persistence of broker data, not auth. But we're working on a bringing auth via a range of databases into VerneMQ, including Redis, Mongo, Postgresql and others.

**Comment by ig0r** [comment]
@larshesel oh, I missed it. thanks for info! It'd be nice to have for both persistence and auth as an option.

Looking forward to it

**Comment by testark** [comment]
@dergraf @ioolkos @larshesel yesterday we experienced major data loss in [this scenario](https://vernemq.com/docs/clustering/#detailed-cluster-leave-case-b-make-a-stopped-node-leave). We are in dire need to move persistence of subscriptions and offline messages from node-local to our PostgreSQL or Redis (or maybe MongoDB) clusters. Is there anything that can be done from your side regarding this?

**Comment by larshesel** [comment]
Hi @testark 

I'm sorry to hear you lost data due to this. What happened? A total node loss?

Swapping out leveldb for another backend is something we want to do, the question is only when it happens. Either it's something we're sponsoring ourselves because we have enough available resources (which we don't at the moment) and we think it's the right issue to tackle (priority) or the work gets sponsored externally or even contributed.

So, at present, we can't say with any confidence when this will happen.

**Comment by larshesel** [comment]
btw @testark if you lost the message due to a plain cluster leave, there should be no problem recovering the messages if you delete the `data/meta/peer_service/cluster_state` file before starting vernemq. Then the node will believe it's a single node and can be rejoined with the rest of the cluster and the queue data should be migrated/merged autmatically including all offline messages.

**Comment by iamyellow** [comment]
hi! for fun, i'd like to dig on this :) it could be pretty amazing be free to choose the backend store for subscriptions and offline queues. recently i'm playing a lot w/ [arb](https://github.com/yinqiwen/ardb) and netflix's [dynomite](https://github.com/Netflix/dynomite), and IMHO i think is an awesome tech, and very easy to work w/ redis protocol, there're good redis drivers out there for elixir/erlang. also one could choose a few backend storage options (those supported in ardb, leveldb is one of them!).
btw, could you tell me which components (or in plain english: erlang apps and files) of the actual vmq do i need to modify in order to achieve that? i understand i need to remove migrations and change the way routing tables work removing node info or make it all pluggable and configurable.

your vmq's plugin system deserves a big kudos :) i think is THE key here, maybe i'm biassed because i'm a big fan of the [plugin design](https://youtu.be/dbB-mICjkQM).

thanks for your time

thanks in advance!
